generated/niscope/session.py

```diff
@@ -1839,6 +1839,127 @@ def configure_vertical(self, range, coupling, offset=0.0, probe_attenuation=1.0,
+
+            fetch_relative_to (enums.FetchRelativeTo): Position to start fetching within one record. If not set, use value of NISCOPE_ATTR_FETCH_RELATIVE_TO
+
+            fetch_offet (int): Offset in samples to start fetching data within each record. The offset is applied relative to fetch_relative_to. The offset can be positive or negative. If not set, use value of NISCOPE_ATTR_FETCH_OFFSET
+
+            fetch_record_number (int): Zero-based index of the first record to fetch.  Use fetch__num_records to set the number of records to fetch. If not set, use value of NISCOPE_ATTR_RECORD_NUMBER
+
+            fetch_num_records (int): Number of records to fetch. Use -1 to fetch all configured records. If not set, use value of NISCOPE_ATTR_NUM_RECORDS
+
+            wfm (list of float): Returns an array whose length is the **numSamples** times number of waveforms. Call _actual_num_wfms to determine the number of waveforms.
+
+
+        Returns:
+            wfm_info (list of WaveformInfo): Returns an array of classed with the following timing and scaling information about each waveform:
+
+                                    -  **relative_initial_x** the time (in seconds) from the trigger to the first sample in the fetched waveform
+                                    -  **absolute_initial_x** timestamp (in seconds) of the first fetched sample. This timestamp is comparable between records and acquisitions; devices that do not support this parameter use 0 for this output.
+                                    -  **x_increment** the time between points in the acquired waveform in seconds -  **actual_samples** the actual number of samples fetched and placed in the waveform array
+                                    -  **gain** the gain factor of the given channel; useful for scaling binary data with the following formula:
+
+                                        .. math::
+
+                                            voltage = binary data * gain factor + offset
+
+                                    -  **offset** the offset factor of the given channel; useful for scaling binary data with the following formula:
+
+                                        .. math::
+
+                                            voltage = binary data * gain factor + offset
+
+                                    Call _actual_num_wfms to determine the size of this array.
+
+        '''
+        import sys
+
+        if num_samples is None and wfm is None:
+            raise TypeError("Either 'num_samples' or 'wfm' must be set")
+
+        # Check optional parameters
+        if fetch_relative_to is not None:
+            self.fetch_relative_to = fetch_relative_to
+        if fetch_offet is not None:
+            self.fetch_offet = fetch_offet
```

Comment:
This is a hidden side effect. It won't be obvious that passing the parameter will change the state of the session.

Ideally, I would get rid of the (public) attributes so that there's only one way to do this. But that cannot be done because of _into flavor. Or can it? We'd have to add them as parameters as well. Then maybe they don't need to all be optional parameters.

Alternatively, the correct solution is for the parameters to not have a visible effect on the Session state. How? I think we'd have to use some sort of property context manager that on __exit__ restores the old value.

What do you think?

Reply:
Also, I don't see where this gets used later on. Do these not have an effect on how you interpret the waveform data?

Reply:
These attributes control how fetch works. I don't think it is a problem that we are setting the attributes on the session since the purpose of this is to simplify the fetch.

Even if we added them to the _into dispatcher function, they would still be optional, it's just they would always be set since the default value wouldn't be `None`, but whatever the default value is, if there is a single value that works. There may not be if there are device specific defaults.

These do not control how I interpret the waveform data, they are used by the driver to determine exactly which samples to fetch.

Reply:
Ok, they'd still be optional.

But the root issue remains, that there is a hidden side effect / WTH factor. A reasonable user wouldn't expect that passing an attribute value would change the state of the Session object.

We can live with it, but I think it's unexpected and thus an issue.

Reply:
I think we are beyond where I have a strong opinion on which way to go with this. I will defer to you or @auchter.

I do agree if we make them private that that should be in a separate PR.

Reply:
I agree with Marcos: I don't like that hidden side effect either.

I lean pretty strongly towards deprecating the fetch attributes in favor of making them parameters. So I think what you have here is a good start, but instead of having these parameters be optional and None, they should be optional and specified to the same defaults that the driver uses. So probably like `def fetch(self, timeout=5.0, samples=1000, records=1, relative_to=enums.FetchRelativeTo.PRETRIGGER, offset=0)`.

I think the goal should be that it's possible to look at a call to fetch and understand, in isolation, what data it's going to retrieve.

Also, looking at that signature... why is `channels` not a parameter to fetch?

Reply:
> ...but instead of having these parameters be optional and None, they should be optional and specified to the same defaults that the driver uses.

What worries me about this is that there is an assumption that the driver will use the same defaults for all devices. This isn't guaranteed, is it? Or perhaps if the assumption is broken and the API has different defaults, it's not a huge deal?

> Also, looking at that signature... why is channels not a parameter to fetch?

Check [the green tip](http://nimi-python.readthedocs.io/en/0.7.0/niscope/functions.html#niscope.Session.fetch). :)
Which I just noticed could use some updating in the text.

Reply:
@texasaggie97 I think, summarizing @auchter's suggestion:

1. All fetch methods should take the parameters.
2. We should make the properties private. This is not something the generator supports yet so it'd be a different PR.
3. The parameters should default to explicit values rather than None for discoverability.
4. The defaults should match driver defaults and hope those never change 🤞

I like that this is an overall simplification of the usage API. And it eliminates the side effect because it's no longer visible to clients. I believe all public fetch methods have some hand-coding so it should not be too bad.